Thumbs/ThumbsMiddleware.cs

```diff
@@ -0,0 +1,122 @@
+﻿using System;
+using System.Threading.Tasks;
+using DLCS.Model.Assets;
+using DLCS.Model.Storage;
+using DLCS.Web.Requests.AssetDelivery;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Http.Extensions;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Logging;
+using Microsoft.Net.Http.Headers;
+using Newtonsoft.Json;
+
+namespace Thumbs
+{
+    public class ThumbsMiddleware
+    {
+        private readonly int cacheSeconds;
+        private readonly IConfiguration configuration;
+        private readonly ILogger<ThumbsMiddleware> logger;
+        private readonly IThumbRepository thumbRepository;
+
+        public ThumbsMiddleware(
+            RequestDelegate next,
+            IConfiguration configuration,
+            ILogger<ThumbsMiddleware> logger,
+            IThumbRepository thumbRepository)
+        {
+            this.configuration = configuration;
+            this.cacheSeconds = configuration.GetValue<int>("responseCacheSeconds", 0);
+            this.logger = logger;
+            this.thumbRepository = thumbRepository;
+        }
+
+        public async Task Invoke(HttpContext context,
```

Comment:
Is it worth having this `public Task Invoke(...` and then omitting the `await` statements inside and instead return the `Task<T>`. E.g. `return RedirectToInfoJson(context);` vs `await RedirectToInfoJson(context);`. 

If performance is key, this would save framework creating a statemachine under the covers - would need tested to verify that it actually makes a difference.

Other than perf the only real difference is that exceptions would be caught by caller, which is framework in this instance so wouldn't make a difference here. However, `await` may make sense if we're wanting to handle exceptions inside this function.

Reply:
This is something we should dissect next week.